hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieHFileReader.java

```diff
@@ -151,15 +154,15 @@ public BloomFilter readBloomFilter() {
   }
 
   public List<Pair<String, R>> readAllRecords(Schema writerSchema, Schema readerSchema) throws IOException {
+    final Option<Schema.Field> keySchemaField = Option.ofNullable(readerSchema.getField(keyField));
     List<Pair<String, R>> recordList = new LinkedList<>();
     try {
       final HFileScanner scanner = reader.getScanner(false, false);
       if (scanner.seekTo()) {
         do {
           Cell c = scanner.getKeyValue();
-          byte[] keyBytes = Arrays.copyOfRange(c.getRowArray(), c.getRowOffset(), c.getRowOffset() + c.getRowLength());
-          R record = getRecordFromCell(c, writerSchema, readerSchema);
-          recordList.add(new Pair<>(new String(keyBytes), record));
+          final Pair<String, R> keyAndRecordPair = getRecordFromCell(c, writerSchema, readerSchema, keySchemaField);
+          recordList.add(new Pair<>(keyAndRecordPair.getFirst(), keyAndRecordPair.getSecond()));
```

Comment:
Instead of allocation a new Pair, cant we simply add keyAndRecordPair to recordList?

Reply:
fixed it.